### PR TITLE
Implement tests on inference scenarios for models

### LINKED ISSUE #1076: Implement tests on inference scenarios for models
### 🚀 Feature Request

We want to have tests that check different inference scenarios for models.

### Proposal

Add tests in `test_inference` on:
1. Forecast on subset of segments:
  * Check that forecast on subset of segments is equal to subset of forecast on all segments.
2. Predict on subset of segments.
  * Check that predict on subset of segments is equal to subset of predict on all segments.
3. Forecast on new segments.
  * Check that forecast fills the target without fail.
4. Predict on new segments.
  * Check that predict fills the target without fail (target will be already filled, but we want to check that it doesn't fail).
5. Predict on future data with gap (we can try to do it like with `forecast`: `TestForecastOutSamplePrefix`, 
`TestForecastOutSampleSuffix`)
  * Check that prediction on prefix of future dataset is the same as prefix of prediction on bigger future dataset.
  * Check that prediction on suffix of future dataset is the same as suffix of prediction on bigger future dataset.

If some tests are failed for some models mark this tests as `to_be_fixed` if we want to fix them. If we understand that it isn't possible to fix this (model's properties) make a separate test case for such models.

We have already tested 
1. How `forecast` works on data in future with gap: `TestForecastOutSampleSuffix`.
3. How `predict` works on data right after train: `TestPredictOutSample`.

### Test cases

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Mr-Geekman**
## Before submitting (must do checklist)

- [x] Did you read the [contribution guide](https://github.com/tinkoff-ai/etna/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs? We use Numpy format for all the methods and classes.
- [x] Did you write any new necessary tests?
- [x] Did you update the [CHANGELOG](https://github.com/tinkoff-ai/etna/blob/master/CHANGELOG.md)?

## Proposed Changes
Look #1076. Also added tests on `MLPModel`.

## Closing issues
Closes #1076.

### CODE REVIEW

File: tests/test_models/test_inference/test_forecast.py [hunk 1]
```diff
@@ -800,3 +836,109 @@ def test_forecast_subset_segments(self, model, transforms, example_tsds):
...
+            (LinearMultiSegmentModel(), [LagTransform(in_column="target", lags=[5, 6])]),
+            (ElasticMultiSegmentModel(), [LagTransform(in_column="target", lags=[5, 6])]),
+            (RNNModel(input_size=1, encoder_length=7, decoder_length=7, trainer_params=dict(max_epochs=1)), []),
+            (
+                MLPModel(input_size=2, hidden_size=[10], decoder_length=7, trainer_params=dict(max_epochs=1)),
+                [LagTransform(in_column="target", lags=[5, 6])],
+            ),
+        ],
+    )
+    def test_forecast_new_segments(self, model, transforms, example_tsds):
```
Comment (alex-hse-repository) [inline tests/test_models/test_inference/test_forecast.py:884]: I guess we can use lags in all the test case to simplify parametrization

Comment (alex-hse-repository) [inline tests/test_models/test_inference/test_forecast.py:884]: May we check somehow that all the multisegment models are here, I mean if we add new model, this test should inform us somehow that we should add this model to it(the same for per-segment models)

File: tests/test_models/test_inference/test_forecast.py [hunk 2]
```diff
@@ -800,3 +836,109 @@ def test_forecast_subset_segments(self, model, transforms, example_tsds):
...
+                        time_varying_known_reals=["time_idx"],
+                        time_varying_unknown_reals=["target"],
+                        static_categoricals=["segment"],
+                        target_normalizer=None,
+                    )
+                ],
+            ),
+        ],
+    )
+    def test_forecast_new_segments_failed_encoding_error(self, model, transforms, example_tsds):
```
Comment (alex-hse-repository) [inline tests/test_models/test_inference/test_forecast.py:918]: Are we going to fix them somehow, as I understand we don't?

File: tests/test_models/test_inference/test_forecast.py [hunk 3]
```diff
@@ -800,3 +836,109 @@ def test_forecast_subset_segments(self, model, transforms, example_tsds):
...
+        with pytest.raises(KeyError, match="Unknown category"):
+            self._test_forecast_new_segments(example_tsds, model, transforms, train_segments=["segment_1"])
+
+    @to_be_fixed(raises=NotImplementedError, match="Per-segment models can't make predictions on new segments")
```
Comment (alex-hse-repository) [inline tests/test_models/test_inference/test_forecast.py:922]: How we are going to fix it?

File: tests/test_models/test_inference/test_forecast.py [hunk 4]
```diff
@@ -558,10 +589,15 @@ def test_forecast_out_sample_suffix(self, model, transforms, example_tsds):
         self._test_forecast_out_sample_suffix(example_tsds, model, transforms)
 
     @to_be_fixed(raises=AssertionError)
+    # Looks like a problem of current implementation of NNs
```
Comment (alex-hse-repository) [inline tests/test_models/test_inference/test_forecast.py:592]: May you create issues for all the cases where we need to fix sth

File: tests/test_models/test_inference/test_predict.py [hunk 5]
```diff
@@ -1,3 +1,5 @@
+from copy import deepcopy
```
Comment (alex-hse-repository) [inline tests/test_models/test_inference/test_predict.py:1]: Looks like some antipattern as we have the set of the same tests but for 2 different methods, may be we can merge them somehow?

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: tests/test_models/test_inference/test_predict.py (hunk 5 above)
Comment (Mr-Geekman) [reply tests/test_models/test_inference/test_predict.py:1]: Why do you think it is the same tests? They are different by its code. Tests that share code I was moved into `common.py` file.

File: tests/test_models/test_inference/test_forecast.py (hunk 3 above)
Comment (Mr-Geekman) [reply tests/test_models/test_inference/test_forecast.py:922]: I don't really know how. We can potentially make it possible for naive models, but for other models I don't think we can do smth like that.

File: tests/test_models/test_inference/test_forecast.py [hunk 6]
```diff
@@ -800,3 +836,109 @@ def test_forecast_subset_segments(self, model, transforms, example_tsds):
...
+
+        # fitting
+        train_ts.fit_transform(transforms)
+        test_ts.transform(train_ts.transforms)
+        model.fit(train_ts)
+
+        # forecasting
+        import torch  # TODO: remove after fix at issue-802
+
+        torch.manual_seed(11)
```
[the remark this answers is not in the document]
Comment (Mr-Geekman) [reply tests/test_models/test_inference/test_forecast.py:862]: We want to set seed before each prediction. This fixture sets seed before the first, but for the second we don't set anything.

File: tests/test_models/test_inference/test_forecast.py (hunk 2 above)
Comment (Mr-Geekman) [reply tests/test_models/test_inference/test_forecast.py:918]: I can mark the as `to_be_fixed`, but I'm not really sure that we will fix this in the near future. As you can see, there are a lot of problems with this models.

Comment (alex-hse-repository) [reply tests/test_models/test_inference/test_forecast.py:918]: May be we can create some categories like "backlog"/"todo" for such cases, just to know that sth can be fixed but we don't have time for this

File: tests/test_models/test_inference/test_forecast.py (hunk 3 above)
Comment (alex-hse-repository) [reply tests/test_models/test_inference/test_forecast.py:922]: Let's try to make it possible for naive models, if don't come to simple solution may be it is better to skip it

**Review by alex-hse-repository — APPROVED** [review]
